### DISCUSSION TITLE
Why do redirects prompt the Invalid URL type error?

### DISCUSSION THREAD

**Discussion by nicolasguasca1** [vercel/next.js#38510] [Help] [answered]
To apply ssl on my local server I set up a custom server on my nextjs app I get this error. Any ideas on how to fix it?

**Comment by icyJoseph** [comment, marked as answer]
What's the hostname looking like? Could you provide more context? Sometimes the localhost resolves to an IP, or IPV6 kind of shape and it might cause some issues, but there's very little info here.

https://nextjs.org/docs/advanced-features/custom-server are you following all of their instructions on this doc?

### Resolution

```ts
const hostname = 'localhost'
const port = 3000
// when using middleware `hostname` and `port` must be provided below
const app = next({ dev, hostname, port })
```

According to the docs,  https://nextjs.org/docs/advanced-features/custom-server, the hostname and port must be passed to the `next` function call.

**Reply by nicolasguasca1** [reply]
Hello @icyJoseph !

Thank you for the reply and excuse my lack of information. I followed this tutorial:

https://blog.wbgneto.com/how-to-set-up-local-ssl-with-nextjs

And this is the middleware.ts file I have:

```import { NextRequest, NextResponse } from "next/server"; 
  
 export const config = { 
   matcher: [ 
     "/", 
     "/([^/.]*)", // exclude `/public` files by matching all paths except for paths containing `.` (e.g. /logo.png) 
     "/site/:path*", 
     "/post/:path*", 
     "/_sites/:path*" 
   ] 
 }; 
  
 export default function middleware(req: NextRequest) { 
   // request url 
   const url = req.nextUrl; 
  
   // Get pathname of request (e.g. /blog-slug) CHANGED DUE TO CANARY VERSION TO WORK BOTH ON DEV AND PROD BUT IT USED TO BE: 
   // const { pathname } = req.nextUrl; 
   const pathname = req.nextUrl.pathname.toString(); 
  
   // Get hostname of request (e.g. demo.vercel.pub) 
   const hostname = req.headers.get("host"); 
  
   if (!hostname) 
     return new Response(null, { 
       status: 400, 
       statusText: "No hostname found in request headers" 
     }); 
  
   const currentHost = 
     process.env.VERCEL_ENV === `production` || 
     process.env.VERCEL_ENV === `preview` 
       ? 
         hostname 
           .replace(`.lomplay.com`, "") 
           // .replace(`.platforms-two.vercel.app`, "") 
           // .replace(`.*.vercel.app`, "") 
           // .replace(regex, "") 
           .replace(`${process.env.VERCEL_URL}`, "") 
           .replace(`${process.env.NEXT_PUBLIC_VERCEL_URL}`, "") 
       : hostname.replace(`.localhost:3000`, ""); 
  
   if (pathname.startsWith(`/_sites`)) 
     return new Response(null, { 
       status: 404 
     }); 
  
   if ( 
     !pathname.includes(".") 
  
   ) { 
     if (currentHost === "tickets") { 
       if ( 
         pathname === "/login" && 
         (req.cookies["next-auth.session-token"] || 
           req.cookies["__Secure-next-auth.session-token"]) 
       ) { 
         url.pathname = "/"; 
         return NextResponse.redirect(url); 
       } 
  
       url.pathname = `/tickets${url.pathname}`; 
       console.log(url); 
       return NextResponse.rewrite(url); 
     } 
   
     url.pathname = `${pathname}`; 
     return NextResponse.rewrite(url); 
   } 
 }
``` 

Do you think there is anything else I need to add for my middleware to work with https?

**Reply by icyJoseph** [reply]
Hi, 

I take it that you are not using 12.2.x? Which version have you got? I can tell you already that in next 12.2.x, you cannot return 404 or 400 directly from middleware.

In the guide you shared, the `hostname`, and `port` are not set, in step 5 specifically, we find this line:

```ts
const dev = process.env.NODE_ENV !== 'production';
const app = next({ dev });
```

However, the documentation clearly states, https://nextjs.org/docs/advanced-features/custom-server:

```ts
const hostname = 'localhost'
const port = 3000
// when using middleware `hostname` and `port` must be provided below
const app = next({ dev, hostname, port })
```

**Reply by nicolasguasca1** [reply]
I am actually using Next 12.2.0 but you just gave me the solution. Thank you so much!

**Reply by icyJoseph** [reply]
No worries! Please don't forget to mark as answered.

**Reply by Aeolun** [reply]
I do not think 'the documentation clearly states' is quite true for something that's hidden in a comment on a code sample. I feel like this should be added to the documentation on middleware.

But thanks for giving me the solution 🙂
